### PR TITLE
refactor(persist): pluggable completion observer

### PR DESCRIPTION
**Pull request by domodwyer**
This PR introduces a new post-persist hook, but uses a no-op observer implementation - there is no functional change in this PR.

---

* refactor(persist): pluggable completion observer (4b3a5c0c2)
      
      Changes the persist system to call into an abstract
      PersistCompletionObserver after the persist task has completed, but
      before releasing the job permit / notifying the enqueuer.
      
      This call happens synchronously, driven by the persist worker to
      completion. A sync construct can easily be made async (by enqueuing work
      into a channel), but not the other way around, so this gives the best
      flexibility.
      
      This trait allows pluggable logic to be inserted into the persist
      system, without tightly coupling it to the implementer's logic (for
      example, replication). One or more observers may be chained together to
      construct an arbitrary sequence of actors.
      
      This commit uses a no-op observer, causing no functional change to the
      system.

### CODE REVIEW

File: ingester2/src/persist/completion_observer.rs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+/// unnecessary slowdown of persistence system.
+///
+/// [`PartitionData::mark_persisted()`]:
+///     crate::buffer_tree::partition::PartitionData::mark_persisted()
+#[async_trait]
+pub(crate) trait PersistCompletionObserver: Send + Sync + Debug {
+    /// Observe the [`CompletedPersist`] notification for the newly persisted
+    /// data.
+    async fn persist_complete(&self, note: Arc<CompletedPersist>);
+}
```
Comment (domodwyer) [inline ingester2/src/persist/completion_observer.rs:22]: This is the abstract post-persist hook.

File: ingester2/src/persist/completion_observer.rs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+#[derive(Debug, Clone)]
+pub struct CompletedPersist {
+    /// The catalog identifiers for the persisted partition.
+    namespace_id: NamespaceId,
+    table_id: TableId,
+    partition_id: PartitionId,
+
+    /// The [`SequenceNumberSet`] of the persisted data.
+    sequence_numbers: SequenceNumberSet,
+}
```
Comment (domodwyer) [inline ingester2/src/persist/completion_observer.rs:34]: This is the information in the completion notification.

File: ingester2/src/persist/completion_observer.rs [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+/// A no-op implementation of the [`PersistCompletionObserver`] trait.
+#[derive(Debug, Default)]
+pub(crate) struct NopObserver;
+
+#[async_trait]
+impl PersistCompletionObserver for NopObserver {
+    async fn persist_complete(&self, _note: Arc<CompletedPersist>) {
+        // the goggles do nothing!
+    }
+}
```
Comment (domodwyer) [inline ingester2/src/persist/completion_observer.rs:82]: This is the no-op implementation that's used in this commit to satisfy the dependencies.

File: ingester2/src/persist/mod.rs [hunk 4]
```diff
@@ -293,6 +306,15 @@ mod tests {
...
             .expect("worker task failed");
 
+        // Assert the notification observer was invoked exactly once, with the
+        // successful persist output.
+        assert_matches!(&completion_observer.calls().as_slice(), &[n] => {
+            assert_eq!(n.namespace_id(), namespace_id);
+            assert_eq!(n.table_id(), table_id);
+            assert_eq!(n.partition_id(), partition_id);
+            assert_eq!(n.sequence_numbers().len(), 1);
+        });
```
Comment (domodwyer) [inline ingester2/src/persist/mod.rs:316]: It does the thing.

File: ingester2/src/persist/completion_observer.rs [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+/// A no-op implementation of the [`PersistCompletionObserver`] trait.
+#[derive(Debug, Default)]
+pub(crate) struct NopObserver;
+
+#[async_trait]
+impl PersistCompletionObserver for NopObserver {
+    async fn persist_complete(&self, _note: Arc<CompletedPersist>) {
+        // the goggles do nothing!
```
Comment (lukebond) [inline ingester2/src/persist/completion_observer.rs:80]: :rofl:

**Review by lukebond — APPROVED** [review]

**Pull request merged**